### PR TITLE
Add loading layer UI indicator

### PR DESCRIPTION
**Pull request by dalthviz**
# Description

Hi, this adds an UI indicator for layers using the `loaded` event. It is part of the implementation of #5326 designs (the implementation of the slow loading screens):

* Adds a loading indicator on the layer item using the [loading.gif](https://github.com/napari/napari/blob/main/napari/resources/loading.gif)
* Changes the layer tooltip text to indicate the loading state if necessary.

## Preview:

* Using the thumbnail space to draw the loading indicator and show a tooltip with a `(loading)` end text:

load

## Some notes:

* To test this you will need to check the `Render Images Asynchronously` option under `File > Preferences > Experimental > Experimental` (otherwise you will not be able to see the loading indicator).
* Uses the [loading.gif](https://github.com/napari/napari/blob/main/napari/resources/loading.gif) resource available as loading indicator (instead of the dots spinner loading icon you can see on the designs at #5326)
* Refactors a little bit the access to the `loading.gif`  file to use a constant

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?
- [x] all tests pass with my change
- [x] I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/_qt/containers/qt_layer_model.py
Comment (andy-sweet) [inline napari/_qt/containers/qt_layer_model.py]: Though I know this is in the designs, seeing the code, I think it's a bit strange that we completely overwrite the tooltip when a layer is loading. `Layer.get_source_str` is still available when a layer is loading. Maybe we should consider something like `'{source} (loading)'.format(source=layer.get_source_str())`?

File: napari/_qt/containers/_layer_delegate.py
Comment (andy-sweet) [inline napari/_qt/containers/_layer_delegate.py]: Based on the videos in the PR description, constantly restarting the loading animation is visually jarring. I'd expect there to be some delay before the animation starts (e.g 1 second). Could we achieve that with a single shot timer? Do we need something more complex to handle multiple calls to start? Or does `QMovie` have some built-in functionality for a delayed start?

**Comment by andy-sweet** [comment]
Thanks for sharing early @dalthviz ! Looks like great progress already. Using the older experimental async setting and `loaded` event makes sense, as you're not blocked by the new approach being integrated. Left some initial comments, but can take a more detailed look later.

File: napari/_qt/containers/_layer_delegate.py
Comment (dalthviz) [reply napari/_qt/containers/_layer_delegate.py]: I'm not sure if `QMovie` has something built-in that could help with this but I will check. Thank you for the feedback!

Comment (brisvag) [reply napari/_qt/containers/_layer_delegate.py]: That was my first reaction as well. I don't knwo what a "good" delay is, but let's give it a go with 1 second and see how it feels when actually interacting with the viewer.

Comment (dalthviz) [reply napari/_qt/containers/_layer_delegate.py]: Checking, instead of stopping the animation we can paused it, which makes things look something like this:

<details>

loading_paused

</details>

Anyhow, I will continue checking for a way to start the animation with some sort of delay

Comment (psobolewskiPhD) [reply napari/_qt/containers/_layer_delegate.py]: Any idea why is the loading dots animation is smooth only when you move around the canvas, else blinky?

Comment (dalthviz) [reply napari/_qt/containers/_layer_delegate.py]: I think is because the indicator is shown/removed when detecting a loaded event and here the layer slice is continuously changing (so `layer.loaded` continuously changes).  Since the slices I'm using in my example don't take to much time to load you get that blinky effect. However, if you interact with the viewer you prevent it from changing the slice so the loaded event is not triggered until you let the viewer continue and show the actual current slice. Checking, adding a delay to show the indicator (so only showing the indicator for slices that take more than the delay to load) makes things look like this:

### Delay 1 second:

delay-1

### Delay - 1.5 seconds:

delay-1-5

Comment (psobolewskiPhD) [reply napari/_qt/containers/_layer_delegate.py]: Thanks for the clear explanation. What got me I guess is that the animation seemed to proceed, rather than restarting per slice.

Comment (dalthviz) [reply napari/_qt/containers/_layer_delegate.py]: Thanks for the question (actually it helped me realize the effect of the delay :+1:)! And so you think it would be better to just restart the animation per slice (as in the initial GIF at https://github.com/napari/napari/pull/5342#issue-1451813387), right?

I'm checking locally things to make the animation less visually jarring and since the animation can be paused I was testing that too, but originally I was thinking in the animation being fully restarted per loaded even. Maybe just adding the delay could be enough?

Comment (psobolewskiPhD) [reply napari/_qt/containers/_layer_delegate.py]: I think with the delay it's more clear what is happening? 🤷‍♂️

Comment (brisvag) [reply napari/_qt/containers/_layer_delegate.py]: What if there was an always-visible indicator that just proceeds along the animation when necessary? I still don't love the jumpy result even after 1.5 delay :/

Comment (dalthviz) [reply napari/_qt/containers/_layer_delegate.py]: That's a really good idea! I think I was able to do that by adding a 100 milliseconds delay to pause the animation instead of delaying 1 second or more the start:

layer_loaded_indicator



What do you think?

Comment (psobolewskiPhD) [reply napari/_qt/containers/_layer_delegate.py]: looks 🔥 to me

**Comment by andy-sweet** [comment]
Just pulled this down and tried it out - worked really well in a bunch of use cases. A layer that is fast to slice showed no loading animation (with experimental async turned on and off), and one that is slow showed the behavior I expected too (only showing the animation while waiting for long loads).

What example are you using? I can't quite get the effect in your videos (where the loading animation shows continuously during playback), so was interested to see if I could reproduce it.

**Comment by dalthviz** [comment]
For testing I have been using the `Open Files as Stack...` option and opening 491 images (each one between 40 and 85 KB).

File: napari/_qt/containers/_layer_delegate.py
Comment (Czaki) [inline napari/_qt/containers/_layer_delegate.py:136]: If I understand correctly, this function is called whenever paint events occur. So if the layer is loaded, every paint event will start a single shot timer. It should be started when movie is already running. 

Also, this looks like something that will be a reason for tests segfaults.

File: napari/_qt/containers/_layer_delegate.py
Comment (brisvag) [reply napari/_qt/containers/_layer_delegate.py]: Yeah, way better, nice!

File: napari/_qt/containers/_layer_delegate.py
Comment (dalthviz) [reply napari/_qt/containers/_layer_delegate.py:136]: Thanks for the feedback @Czaki ! You are correct, this function is called whenever a paint event occur. Regarding when to to start the single shot timer, maybe this definition of the `_paint_loading method` could be better?:

```python
   def _paint_loading(
        self,
        painter: QPainter,
        option: QStyleOptionViewItem,
        index: QtCore.QModelIndex,
    ):
        """Paint loading layer indicator."""
        loaded = index.data(LoadedRole)
        load_rect = option.rect.translated(55, 8)
        h = index.data(Qt.ItemDataRole.SizeHintRole).height() - 16
        load_rect.setWidth(h)
        load_rect.setHeight(h)

        if loaded and self.load_movie.state() == QMovie.Running:
            # Add some delay to check if the layer is still loaded to
            # prevent blinking from the loading indicator
            # when pausing the movie.
            self.layer_index = index
            QTimer.singleShot(100, self._check_loaded)
        elif not loaded:
            self.load_movie.start()

        if self.load_movie.state() == QMovie.Running:
            painter.drawPixmap(load_rect, self.load_movie.currentPixmap())
```

Regarding segfaults, you are right, in fact [I had a test segfaulting before since initialy I used a lambda to call `self._check_loaded` passing to the method the `index` variable](https://github.com/napari/napari/actions/runs/3515853535/jobs/5891642145#step:8:216). However, changing that approach to instead set the `index` as a class attribute (to prevent it from being garbage collected) seems like helped. If you have any further recommendations to be sure we prevent segfaults due to the `QTimer` usage here let me know!

Comment (dalthviz) [reply napari/_qt/containers/_layer_delegate.py:136]: ~I think the latest changes pushed here should prevent segfaults~ Need to further check how to stop the timer. Quite handy to have a test that checks for timers :+1:

Comment (dalthviz) [reply napari/_qt/containers/_layer_delegate.py:136]: Seems like the test are passing now 🎉 (although I had use `QObject.timerEvent` instead of `QTimer.singleShot`)

**Comment by dalthviz** [comment]
Note: Linux/Ubuntu jobs re-runned due to ` E: Failed to fetch https://packages.microsoft.com/ubuntu/22.04/prod/dists/jammy/main/binary-amd64/Packages.bz2  File has unexpected size (67629 != 67496). Mirror sync in progress?` (`setup-qt-libs` action step)

**Comment by andy-sweet** [comment]
@dalthviz : given that you've merged in the new async form `main` I just pulled this down to do some testing, and it seems to be working great!

https://github.com/napari/napari/assets/2608297/0b64a28b-cf5f-4724-8a1f-3d1c4551620b

At first, I though there may be a bug in the new async related to `Layer.loaded` because there's a long loading period (starting ~25s in the video above), but I'm pretty sure that's actually correct and this PR is doing its job of communicating that slicing is still happening.

There are many potential improvements with the new approach to async, but this does a great of covering for us while those improvements come in. What would it take to move this out of draft?

**Comment by andy-sweet** [comment]
I'm still a little unsure about the extra loading icon in the layer list item because that is getting a bit crowded now, but I don't have better strong suggestions. With async slicing off by default, users should never see this, but I think the loading icon will still take up space right?

Replacing the thumbnail might be an option as that is also dependent on slicing, but it might be visually jarring. My sense is that we can go with this for now.

**Comment by dalthviz** [comment]
Hi! So:

> What would it take to move this out of draft?

I put this as draft since I wanted to check locally after the merge if I needed to change something. I did a little change for the timer, but after that things are also working for me locally so I will put it as ready for review :)

Still we need to check if tests pass and also maybe the timer approach I ended up doing here is something to be discussed (`QTimer` vs `QTimerEvent`)

> With async slicing off by default, users should never see this, but I think the loading icon will still take up space right?

Correct, as it is right now, even if the loading indicator is not painted (when async slicing is off) the space for it is kept. In fact, the way the space is achieved is adding some padding to the right to the layer visibility indicator here (while adding also some padding to to the text to keep things symmetric): https://github.com/napari/napari/pull/5342/files#diff-0e3701dcff277586dbc0fc25325132aaf1dd78cc990ce94854121682cec2728f

**Comment by dalthviz** [comment]
As an update, my last commit removes the timer logic and instead queries if the layer slicer async mode is on or if an animation is happening. This prevents the loading indicator from blinking (jumping from loaded to not loaded while the layer slice changes during the layer animation). Maybe the way this info is retrieved should be different, but let me know what do you think or if you have more ideas to improve the visualization of the layer loading!

A preview:

loading_indicator

**Review by jni — APPROVED** [review]
I love @andy-sweet's idea of replacing the thumbnail or even overlaying over previous thumbnail. Like @andy-sweet I'm not super concerned about that happening before merge, I'll leave the decision up to you @dalthviz, and other core devs. Anyway, good stuff! 🙏

**Comment by dalthviz** [comment]
Explored the alternative to use the same space for the loading indicator and the thumbnail and actually I think that it looks nicer :+1:

A preview:

load

File: napari/_qt/containers/qt_layer_list.py
Comment (andy-sweet) [inline napari/_qt/containers/qt_layer_list.py:52]: Without this, I'm assuming the widget wouldn't get a render pass, so we wouldn't see movie animate? Could be worth adding an implementation comment to explain this for future readers.

File: napari/_qt/containers/qt_layer_model.py
Comment (andy-sweet) [inline napari/_qt/containers/qt_layer_model.py]: Prefer to use the public setting instead of the private state for checking is async slicing is enabled. I think the following is equivalent and slightly simpler:

```python
        # Playback with async slicing causes flickering between the thumbnail
        # and loading animation in some cases due quick changes in the loaded state,
        # so report as unloaded in that case to avoid that.
        if get_settings().experimental.async_ and (viewer := current_viewer()):
            viewer_playing = viewer.window._qt_viewer.dims.is_playing
            layer_loaded = layer.loaded and not viewer_playing
```

However, we still need private access for the playing state, which is unfortunate. This maybe suggests the timer logic / throttling might be worth the extra complexity, but I don't feel too strongly yet.

This might be nit-picky, but I would consider moving this special case logic for `loaded` to the `LayerDelegate` because this is a specific workaround for a rendering bug. I feel like the model should just report the underlying value. I can't reproduce similar flickering for the tooltip, so I think that could just use `Layer.loaded` directly in that case.

Feel free to push back here!

**Review by andy-sweet — CHANGES_REQUESTED** [review]
I'd like to see some of the code simplification requested in https://github.com/napari/napari/pull/5342/files#r1244479227 (though moving the code to `LayerDelegate` is optional) but otherwise I'm ready to approve.

Thanks for your effort and patience on this one @dalthviz !

**Comment by jni** [comment]
> Explored the alternative to use the same space for the loading indicator and the thumbnail and actually I think that it looks nicer 👍

Wow! 😍 Yeah I super love it. 😃 Plus it immediately fixes the issue of stale thumbnails. Awesome stuff!

File: napari/_qt/containers/qt_layer_list.py
Comment (dalthviz) [reply napari/_qt/containers/qt_layer_list.py:52]: Checked this and it is mostly needed to be sure that the loading indicator is removed/not painted when appropiate and makes the animation look smoother. Also, while checking this, I noticed a missing call to pause the movie 😅 (otherwise the items will always be asking for a rerender!). Will add a comment and the pause call :+1: Thanks for pointing to check this!

File: napari/_qt/containers/qt_layer_model.py
Comment (dalthviz) [reply napari/_qt/containers/qt_layer_model.py]: Will do the code change, indeed simplier! :+1: Regarding the code placement, I left this logic in the Qt model since I think the delegate should just query the model data, paint it and enable editing it (initial though that made me rethink all the timer logic).

In fact, I was thinking that it could be nice to have some sort of `layer.is_being_processed` attribute that enables to know not only if the layer is loaded but also if some continue action is being done over it, like the animation/successive slicing requests (maybe extending the loaded attribute with logic that can handle such cases could be worthy?).

Regarding the tooltip, I also used the workaround to take into account the playback state since you can end up seeing the `(loading)` text appear and disappear if you move the mouse over the layer item in the list:

load_tooltip

In the end, I also think this logic is kind of a workaround to make the GUI look better, so for sure this is something to keep thinking about and improve!

**Comment by andy-sweet** [comment]
Also tested this with some data from zebrahub

https://github.com/napari/napari/assets/2608297/b99795bc-4ceb-4680-aa7c-7bf75fa2b06d

where this feature makes interaction maybe beareable! Lots of potential for improvement on the rendering side and some implementation details here, but I think this is good enough to approve for now.

I also noticed a potential bug where the loading animation was showing as paused for one layer (see the movie at around ~22s). If you know why that is, then fixing it here would be great. Otherwise, I think we can merge and keep improving as this should only be used when the experimental async flag is on.

**Review by andy-sweet — APPROVED** [review]

File: napari/_qt/containers/qt_layer_model.py
Comment (andy-sweet) [reply napari/_qt/containers/qt_layer_model.py]: Agreed that in the long term the workaround should not be needed at all. I'm wary of adding more rendering state to the layer, but I'm hopeful we can find a good home for something like in the future if needed.

**Comment by andy-sweet** [comment]
> I also noticed a potential bug where the loading animation was showing as paused for one layer (see the movie at around ~22s). If you know why that is, then fixing it here would be great.

I wonder if this is due to the order of operations when updating the layer's loaded state: https://github.com/napari/napari/blob/78b0f3f05a9eb7876f4875ddf01dbd03cfbe9f49/napari/_qt/qt_viewer.py#L599

Always a little unsure when Qt will execute things directly and when it will queue things up, so maybe that explains why the  behavior might be inconsistent?

I think I'm fine moving the thumbnail update after the `loaded` update, since there is a `thumbnail` event that gets emitted when the thumbnail actually changes.

**Comment by dalthviz** [comment]
Checked changing the order of the thumbnail call but that didn't do the trick :/

Seems like what is causing the bug is that one layer is pausing the animation while the other is still loading and since now the loading indicator is paused, no further updates to the viewport are being requested 🤔 Will try to find a way to fix that!

**Comment by psobolewskiPhD** [comment]
Looks pretty awesome and makes me wonder, does this adress https://github.com/napari/napari/issues/786 ?

**Comment by jni** [comment]
I think the answer to your question @psobolewskiPhD is yes???

**Comment by dalthviz** [comment]
I think the latest commit should fix the stuck loading layer indicator :tada:

load

File: napari/_qt/containers/_layer_delegate.py
Comment (andy-sweet) [inline napari/_qt/containers/_layer_delegate.py:160]: Does this imply there is only one shared instance of the `QMovie`? If I remember that's because there's only one instance of this delegate shared across all item views, right?

If that's true, I think it's fine, but you might want to add it to the implementation comment above because it explains the underlying reason why we need this specific logic.

Comment (dalthviz) [reply napari/_qt/containers/_layer_delegate.py:160]: Totally agree! Will expand the comment :+1:

**Comment by andy-sweet** [comment]
Thanks for the updates @dalthviz ! Though there's some future room for improvement, I think this super useful and at least safe for now, so I'll likely merge this either tomorrow or early next week if there are not objections.

**Comment by andy-sweet** [comment]
@dalthviz : can you update the PR description so that it reflects the current state of the changes rather than progress towards that? Notes are useful, but questions and requests for review less so.

**Comment by dalthviz** [comment]
Updated the PR description @andy-sweet ! Removed previous previews, added a better description for the current state preview, and removed questions/pings for review. Feel free to further tweak the description if you think something is missing or needs to be removed :+1:

**Comment by andy-sweet** [comment]
Generally this feature is super useful for me when using async slicing.

But on reflection, I think the timer based / throttled approached is preferred over checking the playing state because it's fairly easy (and fairly common for me) to move the sliders or pan/zoom around an image in a way that makes the loading animation flicker undesirably.

Just wanted to record that somewhere, so I don't forget. No need to do anything yet, especially as this is only available when the async slicing experimental flag is on.

Let's wait and see how useful this is for progressive/tiled rendering and/or if there are design changes that come from that before writing up an issue or putting any effort in.

**Pull request merged**
